_posts/2024-03-19-credential-store-for-passwords.adoc

```diff
@@ -0,0 +1,171 @@
+
+== About Credential Reference
+There are multiple uses for credential stores, but this blog post will dive deeper into using credential-stores to avoid specifying passwords in clear text. Passwords are used for various resources when configuring the WildFly server, such as a `key-store` or a `key-manager`. While it is quick and easy to specify the passwords in clear text, it is not very secure. 
+
+== Creating a Resource with Plaintext Password
+Let us first create a resource with the password specified as plaintext. To do that, let's first start the server: 
+```
+    $ WILDFLY_HOME/bin/standalone.sh 
+```
+Once the server is running, open another terminal and connect to the cli to configure the server: 
+```
+    $ WILDFLY_HOME/bin/jboss-cli.sh --connect
+```
+Now we can create a keystore using a plaintext password: 
+```
+/subsystem=elytron/key-store=serverKS:add(path=server.keystore, relative-to=jboss.server.config.dir, type=PKCS12, credential-reference={clear-text=secret})
+```
+
+When configuring a resource using plaintext password, it is stores in the standalone.xml file as seen below: 
+```
+    <key-store name="serverKS">
+        <credential-reference clear-text="secret"/>
+        <implementation type="PKCS12"/>
+        <file path="server.keystore" relative-to="jboss.server.config.dir"/>
+    </key-store>
+```
+As you can see, the password can easily be obtained from the standalone.xml file or using the read-resource() function on the elytron subsystem: 
+```
+{
+    "outcome" => "success",
+    "result" => {
+        "alias-filter" => undefined,
+        "credential-reference" => {"clear-text" => "secret"},
+        "path" => "server.keystore",
+        "relative-to" => "jboss.server.config.dir",
+        "required" => false,
+        "provider-name" => undefined,
+        "providers" => undefined,
+        "type" => "PKCS12"
+    }
+}
+```
+However, this can be changed using a credential-store and alias to point to the password instead. 
+
+== Create a Credential Store
+A credential store can hold multiple passwords at once, with each password uniquely identified by an alias. When we want to use a password for a resource, we can specify which credential-store it is in and which alias it is under. Let's first create a credential store: 
+
+```
+/subsystem=elytron/credential-store=myCredStore:add(location=mycredstore.cs, relative-to=jboss.server.config.dir, credential-reference={clear-text=StorePassword}, create=true)
+```
+
+If you navigate to WILDFLY_HOME/standalone/configuration, you will see a new file has been created there named mycredstore.cs. This file is used to store all the credentials in a credential-store. If you try to open it using Vim or another file viewer, you will see that the file is not human readable. As a result, the passwords are secured. It is possible to programmatically read the passwords, which is what WildFly does when dereferencing the credential reference to access a resource. 
```

Comment:
Just a minor suggestion, if you have the StorePassword (that was specified in clear text) then it is not secured just because it is human readable, and wildfly dereferencing is public code, so I would just remove the sentence: `As a result, the passwords are secured. ` and leave the rest:

If you navigate to WILDFLY_HOME/standalone/configuration, you will see a new file has been created there named mycredstore.cs. This file is used to store all the credentials in a credential-store. If you try to open it using Vim or another file viewer, you will see that the file is not human readable. It is possible to programmatically read the passwords, which is what WildFly does when dereferencing the credential reference to access a resource.

Reply:
fixed!